relay-server/src/services/processor/dynamic_sampling.rs

```diff
@@ -62,43 +62,40 @@ pub fn ensure_dsc(state: &mut ProcessEnvelopeState<TransactionGroup>) {
 }
 
 /// Computes the sampling decision on the incoming event
-pub fn run(state: &mut ProcessEnvelopeState<TransactionGroup>, config: &Config) -> SamplingResult {
+pub fn run<G>(state: &mut ProcessEnvelopeState<G>, config: &Config) -> SamplingResult
+where
+    G: Sampling,
+{
     // Running dynamic sampling involves either:
     // - Tagging whether an incoming error has a sampled trace connected to it.
     // - Computing the actual sampling decision on an incoming transaction.
-    match state.event_type().unwrap_or_default() {
-        EventType::Default | EventType::Error => {
-            tag_error_with_sampling_decision(state, config);
-            SamplingResult::Pending
-        }
-        EventType::Transaction => {
-            match state.project_state.config.transaction_metrics {
-                Some(ErrorBoundary::Ok(ref c)) if c.is_enabled() => (),
-                _ => return SamplingResult::Pending,
-            }
 
-            let sampling_config = match state.project_state.config.sampling {
-                Some(ErrorBoundary::Ok(ref config)) if !config.unsupported() => Some(config),
-                _ => None,
-            };
+    // For transactions, we require transaction metrics to be enabled before sampling.
+    let transaction_metrics_enabled = matches!(&state.project_state.config.transaction_metrics, Some(ErrorBoundary::Ok(c)) if c.is_enabled());
+    let group = state.managed_envelope.group();
+    if matches!(group, ProcessingGroup::Transaction) && !transaction_metrics_enabled {
+        return SamplingResult::Pending;
+    }
```

Comment:
We have a sampling trait now, this would be something the trait should encapsulate.

It's now generic what you can pass in here, but we go back from generic to specific again in the implementation. Maybe not having a trait in this case is better, then it won't give the illusion that the function does the right thing if you just implement the trait.